File: frontend/src/api/integrations/slack.ts
const API = import.meta.env.VITE_API_URL || "";
import { apiFetch } from "../http";

export interface SlackStatus {
  has_token: boolean;
  token_masked: string | null;
  channel_id: string | null;
  channel_name: string | null;
  workspace_name: string | null;
  workspace_url: string | null;
  enabled: boolean;
  auto_push: boolean;
  include_media: boolean;
  include_message_links: boolean;
  mention_channel: boolean;
  is_configured: boolean;
}

export interface SlackChannel {
  id: string;
  name: string;
  is_private: boolean;
  num_members: number | null;
}

export interface SlackCredentials {
  bot_token?: string;
}

export interface SlackTestResult {
  success: boolean;
  message: string;
  workspace_name?: string | null;
  workspace_url?: string | null;
  bot_name?: string | null;
}

async function parseError(r: Response): Promise<string> {
  try {
    const d = await r.json();
    return (d as { detail?: string }).detail || r.statusText;
  } catch {
    return r.statusText;
  }
}

export async function fetchSlackStatus(): Promise<SlackStatus> {
  const r = await apiFetch(`${API}/api/integrations/slack/status`);
  if (!r.ok) throw new Error(await parseError(r));
  return r.json();
}

export async function saveSlackSettings(body: Record<string, unknown>): Promise<SlackStatus> {
  const r = await apiFetch(`${API}/api/integrations/slack/settings`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error(await parseError(r));
  return r.json();
}

export async function testSlackConnection(creds?: SlackCredentials): Promise<SlackTestResult> {
  const r = await apiFetch(`${API}/api/integrations/slack/test`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(creds || {}),
  });
  if (!r.ok) throw new Error(await parseError(r));
  return r.json();
}

export async function fetchSlackChannels(creds?: SlackCredentials): Promise<SlackChannel[]> {
  const r = await apiFetch(`${API}/api/integrations/slack/channels`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(creds || {}),
  });
  if (!r.ok) throw new Error(await parseError(r));
  return r.json();
}

File: backend/app/schemas/slack.py
from pydantic import BaseModel


class SlackStatusOut(BaseModel):
    has_token: bool = False
    token_masked: str | None = None
    channel_id: str | None = None
    channel_name: str | None = None
    workspace_name: str | None = None
    workspace_url: str | None = None
    enabled: bool = False
    auto_push: bool = False
    include_media: bool = True
    include_message_links: bool = True
    mention_channel: bool = False
    is_configured: bool = False


class SlackSettingsIn(BaseModel):
    bot_token: str | None = None
    clear_token: bool = False
    channel_id: str | None = None
    channel_name: str | None = None
    workspace_name: str | None = None
    workspace_url: str | None = None
    enabled: bool | None = None
    auto_push: bool | None = None
    include_media: bool | None = None
    include_message_links: bool | None = None
    mention_channel: bool | None = None


class SlackCredentialsIn(BaseModel):
    bot_token: str | None = None


class SlackTestIn(SlackCredentialsIn):
    pass


class SlackTestOut(BaseModel):
    success: bool
    message: str
    workspace_name: str | None = None
    workspace_url: str | None = None
    bot_name: str | None = None


class SlackChannelOut(BaseModel):
    id: str
    name: str
    is_private: bool = False
    num_members: int | None = None
